From 73ab6fa258ae4a730b0299b76c3e0bf31665d776 Mon Sep 17 00:00:00 2001
From: Leonard Crestez <leonard.crestez@nxp.com>
Date: Wed, 18 Jul 2018 16:14:34 +0300
Subject: [PATCH 4169/5242] MLK-18918: dma: imx-sdma: Mark fw_loaded false on
 resume

commit  6347aab298b58f8c15bc6dde601cc2d7c821c205 from
https://source.codeaurora.org/external/imx/linux-imx.git

When the sdma engine is suspended firmware might get lost and need to be
reloaded. Right now fw_loaded is set to false in suspend code and SDMA
is disabled until FW is reloaded on resume.

However when entering standby mode the firmware is not lost, the resume
code detects this situation and skips reloading firmware and restoring
contexts. Unfortunately this incorrectly leaves fw_loaded set to false
so SDMA is unusable after a standby.

Fix by setting fw_loaded to false at the start of sdma_resume instead
of the end of sdma_suspend. This is the point at which we know that
firmware was lost and needs reloading.

Fixes: c5e50134e7b1 ("MLK-18315-1 dma: imx-sdma: add fw_loaded check")

Signed-off-by: Leonard Crestez <leonard.crestez@nxp.com>
Acked-by: Fugang Duan <fugang.duan@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/imx-sdma.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/drivers/dma/imx-sdma.c b/drivers/dma/imx-sdma.c
index decbb05..1633ad9 100644
--- a/drivers/dma/imx-sdma.c
+++ b/drivers/dma/imx-sdma.c
@@ -2534,8 +2534,6 @@ static int sdma_suspend(struct device *dev)
 			sdma->save_regs[i] = readl_relaxed(sdma->regs + 4 * i);
 	}
 
-	sdma->fw_loaded = false;
-
 	clk_disable(sdma->clk_ipg);
 	clk_disable(sdma->clk_ahb);
 
@@ -2562,6 +2560,8 @@ static int sdma_resume(struct device *dev)
 		return 0;
 	}
 
+	/* Firmware was lost, mark as "not ready" */
+	sdma->fw_loaded = false;
 	sdma->suspend_off = true;
 
 	/* restore regs and load firmware */
-- 
1.7.9.5

